server/raft.go

```diff
@@ -1160,13 +1161,40 @@ func (n *raft) isCatchingUp() bool {
 	return n.catchup != nil
 }
 
-// Lock should be held.
+// Lock should be held. This function may block for up to ~5ms to check
+// forward progress in some cases.
 func (n *raft) isCurrent() bool {
-	// First check if have had activity and we match commit and applied.
-	if n.commit == 0 || n.commit != n.applied {
-		n.debug("Not current, commit %d != applied %d", n.commit, n.applied)
-		return false
+	// First check if have had activity and that commit and applied are equal.
+	// If not, wait for a short period of time and see if we are making forward
+	// progress.
+	startCommit, startDelta := n.commit, n.commit-n.applied
+	if startCommit == 0 || startDelta > 0 {
+		var progress bool
+		n.RUnlock()
+		for i := 0; i < 10; i++ { // 5ms, in 0.5ms increments
+			time.Sleep(time.Millisecond / 2)
+			n.RLock()
```

Comment:
Let's just simplify locking and put unlock just around the sleep.

n.RUnlock()
time.Sleep()
n.RLock()

We also need to make sure we are holding the RLock and not Lock..

Reply:
You're right, that's simpler — done.

Reply:
Are we guaranteed to only be holding the RLock?

Reply:
Both the call sites for `isCurrent` are taking out RLocks only.